Improvements to the documentation for v0.7.0 (#213)
ds4dm/ecole: docs/discussion/gym-differences.rst

In :py:meth:`~ecole.environment.Environment.reset` a ``reward_offset`` is returned.
This is not only a difference with OpenAI Gym, but also with the MDP formulation.
Its purpose is not to provide additional input to the learning algorithms, but rather to help
researchers better benchmark the resulting performance.
Indeed, :py:class:`~ecole.typing.RewardFunction` are often designed so that their cumulative sum match a
metric on the terminal state, such as solving time or number of LP iterations: this is because final metrics 
are often all that matter.
However, for learning, a single reward on the terminal state is hard to learn from.
It is then divided over all intermediate transitions in the episode.

Rather than providing a different mean of evaluating such metrics, we chose to reuse the
environments to compute the cummulative sum, and therfore need the ``reward_offset`` to exactly
match the metric.

No observation on terminal states
---------------------------------
On terminal states, in OpenAI Gym as in Ecole, no further action can be taken and the environment
needs to be :py:meth:`~ecole.environment.Environment.reset`. In Ecole, when an episode is over (that is, when 

> the ``done`` flag is ``True``), environments return ``None`` as observation. This is in contrast with OpenAI Gym,

Reviewer: ```suggestion
the ``done`` flag is ``True``), environments always return ``None`` as the observation. This is in contrast with OpenAI Gym,
```

where some environments do return observations on terminal states.

This can be explained as follows: most of the time, a terminal state in Ecole is a solved problem.
This means that some complex observations cannot be extracted because they require information that
simply does not exist.
For instance, the :py:class:`~ecole.observation.NodeBipartite` observation function extracts some 
information about the LP solution of the current branch-and-bound node.
When the problem is solved, for example on a terminal state of the
:py:class:`~ecole.environment.Branching` environment, there might not be a current node, or a linear 
relaxation problem, from which this information can be extracted. For these reasons, one would find a 
``None`` instead of an observation on terminal states.

> In any case, one might note that in reinforcement learning, the observation of a terminal state is actually never used.

Reviewer: ```suggestion
In any case, one might note that in reinforcement learning, the observation of a terminal state is usually not very useful.
```

It is not given to a policy to take the next action (because there are not any), and hence never

> used for learning either, so the difference has no impact in practice.

Reviewer: ```suggestion
used for learning either, so not returning a final observation has no impact in practice.
```
